### PR TITLE
feat(common): make CompletionQueueImpl mockable [2]

### PR DESCRIPTION
**Pull request by coryan**
Made `CompletionQueueImpl` a pure abstract class, that makes it fully
mockable. Move the implementation to `DefaultCompletionQueueImpl`. I had
to move some other classes around to compile the code and avoid strange
helpers in `completion_queue_impl.h`.

Part of the work for #4718

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5039)

### CODE REVIEW

File: google/cloud/internal/default_completion_queue_impl.h [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  /// Return the asynchronous operation associated with @p tag.
+  std::shared_ptr<AsyncGrpcOperation> FindOperation(void* tag);
+
+  /// Unregister @p tag from pending operations.
+  void ForgetOperation(void* tag);
+
+  grpc::CompletionQueue cq_;
+  mutable std::mutex mu_;
+  bool shutdown_{false};  // GUARDED_BY(mu_)
+  std::unordered_map<std::intptr_t, std::shared_ptr<AsyncGrpcOperation>>
```
Comment (devbww) [inline google/cloud/internal/default_completion_queue_impl.h:68]: Cannot we use the `void*` directly as the key, and avoid the `reinterpret_cast<>`s?

File: google/cloud/internal/completion_queue_impl.h [hunk 2]
```diff
@@ -17,281 +17,57 @@
...
 /**
  * The implementation details for `CompletionQueue`.
  *
  * `CompletionQueue` is implemented using the PImpl idiom:
  *     https://en.wikipedia.org/wiki/Opaque_pointer
- * This is the implementation class in that idiom.
+ *
+ * This class defines the interface for the Impl class in that idiom.
  */
 class CompletionQueueImpl {
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/completion_queue_impl.h:45]: A bit... less weird when using the PImpl idiom, :shrug:   I am open to other names (this is in `internal` anyways), though I would get easily confused if we called it  `google::cloud::internal::CompletionQueue` (cf. `google::cloud:CompletionQueue` and `grpc::CompletionQueue`).

File: google/cloud/testing_util/fake_completion_queue_impl.cc [hunk 3]
```diff
@@ -64,18 +64,72 @@ class FakeAsyncFunction : public internal::AsyncGrpcOperation {
...
+}
+
+void FakeCompletionQueueImpl::Shutdown() {
+  std::unique_lock<std::mutex> lk(mu_);
+  shutdown_ = true;
+  while (!pending_ops_.empty()) {
+    auto op = std::move(pending_ops_.back());
+    pending_ops_.pop_back();
+    lk.unlock();
+    op->Notify(false);
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/testing_util/fake_completion_queue_impl.cc:79]: Good catch.  The "non-fake" version of this class stops accepting new operations once `shutdown_` is true, we need to do the same here and I had missed a couple of places.

File: google/cloud/internal/completion_queue_impl.cc [hunk 4]
```diff
@@ -17,246 +17,10 @@
...
-    void* tag;
-    bool async_next_ok;
-    auto deadline =
-        std::chrono::system_clock::now() + std::chrono::milliseconds(1);
-    status = cq_.AsyncNext(&tag, &async_next_ok, deadline);
-  } while (status == grpc::CompletionQueue::GOT_EVENT);
-}
-
-}  // namespace internal
+namespace internal {}  // namespace internal
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/completion_queue_impl.cc]: Removed.

File: google/cloud/async_operation.h [hunk 5]
```diff
@@ -48,6 +48,36 @@ class AsyncOperation {
   virtual void Cancel() = 0;
 };
 
+namespace internal {
+
+/**
+ * Represents an AsyncOperation which gRPC understands.
+ *
+ * When applications create an asynchronous operations with a `CompletionQueue`
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/async_operation.h]: Done.

File: google/cloud/internal/default_completion_queue_impl.h (hunk 1 above)
Comment (coryan) [reply google/cloud/internal/default_completion_queue_impl.h]: Done.

File: google/cloud/internal/completion_queue_impl.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (devbww) [reply google/cloud/internal/completion_queue_impl.h:45]: I see ... this is the interface for a set of classes that happen to have "Impl" in their name (i.e., those classes are "Impl" implementations).

**Review by devbww — APPROVED** [review]

**Pull request merged**